src/adapter/src/catalog.rs

```diff
@@ -6184,6 +6398,53 @@ impl Catalog {
                     ))
                 .collect();
         }
+
+        *privileges = MzAclItem::group_by_grantee(flat_privileges);
+    }
+
+    fn grant_object_privilege(
+        privileges: &mut BTreeMap<RoleId, Vec<MzAclItem>>,
+        privilege: MzAclItem,
+    ) {
+        let grantee_privileges = privileges.entry(privilege.grantee).or_default();
+        if let Some(existing_privilege) = grantee_privileges
+            .iter_mut()
+            .find(|cur_privilege| cur_privilege.grantor == privilege.grantor)
+        {
+            // sanity check that the catalog data is consistent.
+            assert_eq!(
+                privilege.grantee, existing_privilege.grantee,
+                "catalog privileges out of sync"
+            );
+            existing_privilege.acl_mode = existing_privilege.acl_mode.union(privilege.acl_mode);
+        } else {
+            grantee_privileges.push(privilege);
+        }
+    }
+
+    fn revoke_object_privilege(
+        privileges: &mut BTreeMap<RoleId, Vec<MzAclItem>>,
+        privilege: MzAclItem,
+    ) {
+        let grantee_privileges = privileges.entry(privilege.grantee).or_default();
+        if let Some(existing_privilege) = grantee_privileges
+            .iter_mut()
+            .find(|cur_privilege| cur_privilege.grantor == privilege.grantor)
+        {
+            // sanity check that the catalog data is consistent.
+            assert_eq!(
+                privilege.grantee, existing_privilege.grantee,
+                "catalog privileges out of sync"
+            );
+            existing_privilege.acl_mode =
+                existing_privilege.acl_mode.difference(privilege.acl_mode);
+        }
```

Comment:
Don't need an else block here like above because it's not a problem to attempt to revoke a privilege that already doesn't exist?

Reply:
Yeah, that just ends up as a no-op.